cowprotocol/cowswap
# chore: add abis library (+fix e2e, cosmos, unit test)

# Summary
Fixes some issues with cypress, cosmos, and local development related to a failed import of the ABIs.

Additionally, improves how we handle ABIs (adds a new library for them).

## New library
This simplifies the use of our ABIs.

<img width="842" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/d4011205-141a-4f0c-8a91-3d41f670a1d3">



From CoW Swap point of view  all Abi types, json, etc will be coming from a new library:
`import { Erc20, GPv2Settlement, CoWSwapEthFlow } from '@cowprotocol/abis'`

It doesn't matter any more if they are legacy types or not, now all relevant ABIs are in a new library.

## Generation of types
It was a bit annoying to generate the types over and over (on every install). 

When I did the PR, i actually respected this behavior. However, later I decided to commit the generated types into the project, reason is:
- They ABIs don't change often, so we don't need to re-generate often
- Re-generating is still simple to do: `nx abis abis`
- We save precious time for each CI build and every time we don `yarn install`


All generated types are added in `src/generated`

<img width="490" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/bb9e06c2-c939-4532-bd10-3fe9839d1fa9">


# v3 types

Initially I was generating Uniswap v3 types, however, I realised they were generating a lot of files which we don't need. Only one ABI  (and their types) was used, so I brought it to `legacy` and re-exported it.

Now we don't need to generated uni-v3 types anymore (which was a bit slow)

# Cypress work!!!
<img width="645" alt="image" src="https://github.com/cowprotocol/cowswap/assets/2352112/623356c6-e90b-43c9-ad1e-daafed0b9df5">


# ~Not in this PR (will fix in another, or HELP WANTED )~ --> Thanks Tukan!
> 🚀 EDIT: Tukan found a solution with #2754 , so now it works also the unit tests

~I had to cheat to make test work. I tried to fix them but I was not able.~
~In principle, its all good, but there must be sth with my craco-jest config~! 

~The problem is that absolute import such us `import { Erc20 } from '@cowprotocol/abis` don't work any more in the test (in the build and webserver works.~

~These two changes should make it work~!
- https://github.com/cowprotocol/cowswap/blob/2af2a1538818bd2f724f9e7e5ad9f1f08861faaa/craco.config.js#L67
- https://github.com/cowprotocol/cowswap/blob/2af2a1538818bd2f724f9e7e5ad9f1f08861faaa/tsconfig.base.json#L31

~They don't~ 😭

~I disabled the tests so they would pass~! 
~Does anyone have a happy idea~?


 # To Test

- Make sure test pass
- Make sure e2e pass
- Make sure you can generate the types: `nx abis abis`
- Make sure comsos work
- Make sure CoW Swap works!
- Read README of the new library
